curp/src/server/curp_node.rs

```diff
@@ -354,60 +370,135 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
     #[allow(clippy::integer_arithmetic)]
-    async fn sync_followers_daemon(
-        curp: Arc<RawCurp<C, RC>>,
-        connects: HashMap<u64, Arc<impl InnerConnectApi + ?Sized>>,
-        shutdown_trigger: shutdown::Trigger,
-    ) {
-        let mut shutdown_listener = shutdown_trigger.subscribe();
+    async fn sync_followers_daemon(curp: Arc<RawCurp<C, RC>>) {
+        let mut shutdown_listener = curp.shutdown_listener();
         let leader_event = curp.leader_event();
         loop {
             if !curp.is_leader() {
-                shutdown_trigger.mark_sync_daemon_shutdown();
-                shutdown_trigger.check_and_shutdown();
+                curp.shutdown_trigger().mark_sync_daemon_shutdown();
+                curp.shutdown_trigger().check_and_shutdown();
                 tokio::select! {
                     _ = shutdown_listener.wait_self_shutdown() => {
                         debug!("sync follower daemon exits");
                         return;
                     }
                     _ = leader_event.listen() => {
-                        shutdown_trigger.reset_sync_daemon_shutdown();
+                        curp.shutdown_trigger().reset_sync_daemon_shutdown();
                     }
                 }
             }
-            let futs = connects.iter().map(|(id, connect)| {
-                let sync_event = curp.sync_event(*id);
-                Self::sync_follower_task(
+            let (sync_task_tx, mut sync_task_rx) = mpsc::channel(128);
+            let mut remove_events = HashMap::new();
+
+            for c in curp.connects().iter() {
+                let sync_event = curp.sync_event(c.id());
+                let remove_event = Arc::new(Event::new());
+                let fut = Self::sync_follower_task(
                     Arc::clone(&curp),
-                    Arc::clone(connect),
+                    c.value().clone(),
                     sync_event,
-                    shutdown_listener.clone(),
-                )
-            });
-            _ = futures::future::join_all(futs).await;
+                    Arc::clone(&remove_event),
+                );
+                sync_task_tx
+                    .send(tokio::spawn(fut))
+                    .await
+                    .unwrap_or_else(|_e| unreachable!("receiver should not be closed"));
+                _ = remove_events.insert(c.id(), remove_event);
+            }
+
+            let _handle = tokio::spawn(Self::conf_change_handler(
+                Arc::clone(&curp),
+                remove_events,
+                sync_task_tx,
+            ));
+
+            while let Some(h) = sync_task_rx.recv().await {
+                let leader_retired = h
+                    .await
+                    .unwrap_or_else(|e| unreachable!("sync follower task panicked: {e}"));
+                if leader_retired {
+                    break;
+                }
+            }
```

Comment:
Since the sync task is mutable, we can't join all task handles at the same time. In the current implementation, whenever the current handle of sync task returns true, it will break and not wait for another task handle. It may not be the fastest, but it doesn't require block too long.